### PR TITLE
Add EIP-6051: Private Key Encapsulation

### PR DESCRIPTION
**Pull request by weiji-cryptonatty**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:9]: This seems an `Interface` category

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:16]: [Editorial] Suggest spell out `ECIES` what it's abbreviated for and provide relevant brief explanation

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:51]:
```suggestion
7.  `cipher  := authenticated_encryption(symAlg,  skey,  IV, data=sk)`. The symmetric cipher algorithm `symAlg` and authentication scheme are decided by the version parameter. No additional authentication data `aad` is used.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:73]: Suggest putting in the response format here

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:70]: Specify the length of version and signerPubKey here or mention it will be specified below.

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:9]: I double checked eip-1 for `Interface` use. I don't think this should be an `Interface` category. The requests are optional for wallet applications. It does _not_ require _all_ wallet applications to implement. Also, EIP-5630 uses `ERC` category as well, while it also provides eth_ api additions like `eth_getEncryptionPublicKey ` and `eth_decrypt `.

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:16]: added

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:70]: moved "Encoding of data and message" prior to "Requests".

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:73]: Added. The `Requests` here are calls to window.ethereum object and then gets a return value as in a javascript function call:

```javascript
window.ethereum.request({
	method: 'eth_generateEphemeralKeyPair',
	params: [version, signerPubKey],
})
// expected return value: R
```

Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:73]: Also the test vectors shall clarify further.

**Comment by weiji-cryptonatty** [comment]
Hi @xinbenlv I had pushed some more commits to address all issues except the `ERC` vs `Interface`. Could you please review again? Thanks a lot.

**Comment by xinbenlv** [comment]
Glad to hear! 

I will continue to review the remaining 2/3 of the EIPs.

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:27]: So far Motivation is good enough for merge as a `Draft`.

Down the road, I think this EIP could still benefit a more thorough and elaborate demonstrated of use cases. Such as the scenarios when a cold wallet exports some private keys and to which third party. This helps reader understand the security model this EIP tries to address.

**Comment by xinbenlv** [comment]
I am going to mainly focus on the editorial review for now, so as we could unblock it for merging as `draft`, but I do have a few technical questions / feedback when conducting my technical peer review, and they will come later

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:123]: For R3, in your typical use case, who is the "requester" and who is the "responder", and is the communication channel between them considered "secured?

Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:123]: You may think of it as a function call. All requests here (R1, R2, R3) are supposed to be called within a web page:

```javascript
window.ethereum.request){
...})
```

Wallets like MetaMask, Tally Ho  or any WalletConnect apps could be called up to process the request (I am not pretty sure about WalletConnect). A hardware/cold wallet may be working behind, connected in a secure manner.

Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:123]: Regarding the security or threat model, we need to prevent a malicious party from intercepting or tampering with the Recipient public key `R`, therefore the Sender application may reject the request R2 if no signature is provided, or not provided as expected (`signerPubKey` is not signed by `trustedPubKey`, just for example). But signature is not mandatory. And the intention is to provide implementation flexibility. For example, a messaging application as Recipient may not have a proper signature, in such cases, it is expected that the private key being encapsulated is dedicated for messaging purposes.

It is possible that Sender application (under attacker's control) provides a 'fake' or 'wrong' private key in R2, but then in R3 user will find out by checking out the returned address of the intake key. A private key installed by attacker may be used to decrypt confidential information or steal funds. This is unlikely, though, as if Sender application is already under attacker's control, user funds have already been compromised.

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [inline EIPS/eip-6051.md:23]: Reviewing the motivation, however, it seems in this exact case we may have alternative request to R2 and R3, that is, each party $P_i$ in a MPC setting may generate its own ephemeral key pair $(r_i, R_i)$, and then Sender application may safely split `sk` safely into several pieces $sk_i$, then encapsulate each $sk_i$ to $R_i$. This effectively replaces the usual Distributed Key Generation procedure in a typical MPC protocol. 

To be fair, this may be added as an amendment to EIP-6051. I need to survey MPC protocols to see if the above suggestions work in their security models.

**Comment by weiji-cryptonatty** [comment]
> I am going to mainly focus on the editorial review for now, so as we could unblock it for merging as `draft`, but I do have a few technical questions / feedback when conducting my technical peer review, and they will come later

Hi @xinbenlv any updates to the editorial review? Any outstanding issues? How can I be of assistance to help you help me? 

Also you are welcome to post any technical questions or feedback to the discussion thread. Many thanks.

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:20]:
```suggestion
There are various cases in which we might want to export one of many private keys from a much more secure but less convenient wallet, which is controlled with a seed or passphrase.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:16]:
```suggestion
This EIP proposes a mechanism to encapsulate a private key so that it could be securely relocated to another application without providing the seed. This EIP combines `ECIES` (Elliptic Curve Integrated Encryption Scheme) and optional signature verification under various choices to ensure that the private key is encapsulated by a known or trusted party.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:22]:
```suggestion
1.  We might dedicate one of many private keys for messaging purposes, and that private key is probably managed in a not-so-secure manner;
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:33]:
```suggestion
We hereby define:
- *Sender* as the party who holds in custody the private key to be encapsulated. *Sender Application* as the client-side application that said Sender uses.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:35]:
```suggestion
- *Recipient* as the party who accepts the encapsulated private key, unwrap and then use it. *Recipient Application* as the client-side application that Recipient uses to receive the encapsulated private key.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:39]:
```suggestion
The basic idea is to encapsulate the private key with ECIES. To ensure that the ephemeral public key to encapsulate the private key is indeed generated from a trusted party and has not been tampered with, we also provided an option to sign that ephemeral public key in this protocol.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:41]:
```suggestion
There should be a mandatory  `version`  parameter. This allows various kinds of Key Encapsulation Mechanisms to be adopted depending on the security considerations or preferences. The list shall be short to minimize compatibility issues among different vendors.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:43]:
```suggestion
In addition to a `version` parameter, the following keys and functions are involved:
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:84]:
```suggestion
`signerPubKey`  is optional. If provided, it is assumed that the implementation has the corresponding private key and the implementation MUST sign the ephemeral public key (in the form of what is to be returned). The signature algorithm is determined by the curve part of the  `version`  parameter, that is, ECDSA for secp256k1, and Ed25519 for Curve25519. And in this situation, it should be the case that the sender trusts  `signerPubKey`, no matter how this trust is maintained. If not, the next request WILL be rejected by Sender application. Also see **Security Considerations**.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:115]:
```suggestion
If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format etc. SHALL fail the call and result in empty return and optional error logs.
```
```suggestion
If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format MUST either fail the call or result in empty return and optional error logs.
```

I am not sure if "SHALL fail the call or return error with error logs" is logically sound here. Changing "MUST either fail the call, or return empty with error logs"

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:167]:
```suggestion
A critical difference of this [EIP-6051](./eip-6051.md) with [EIP-5630](./eip-5630.md) is that, as the purpose of key encapsulation is to transport a private key securely, the public key from the key-recipient should be ephemeral, and mostly used only one-time. While in EIP-5630 settings, the public key of the message recipient shall be stable for a while so that message senders can encrypt messages without key discovery every time.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:169]:
```suggestion
There is security implication to this difference, including perfect forward secrecy. We aim to achieve perfect forward secrecy by generating ephemeral key pairs in both sides every time: 

1) first the recipient shall generate an ephemeral key pair, retain the private key securely, and export the public key;
2) then the key sender securely wrap the private key in ECIES, with another ephemeral key pair, then destroy the ephemeral key securely;
3) finally the recipient can unwrap the private key, then destroy its ephemeral key pair securely. After these steps, the cipher text in transport intercepted by a malicious 3rd party, is no longer decrypt-able.
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:251]:
```suggestion
Then it proceeds to verify the second layer signature, also as ECDSA over secp256k1:
```

File: EIPS/eip-6051.md
Comment (xinbenlv) [inline EIPS/eip-6051.md:456]:
```suggestion
1. We are not considering post-quantum security. If the quantum computer becomes a materialized threat, the underlying cipher of Ethereum and other L1 chains would have been replaced, and this EIP will be outdated then (as the EC part of ECIES is also broken).
```

Comment (xinbenlv) [reply EIPS/eip-6051.md:456]: I think this is generally omittable because the whole Ethereum rely on ECDSA / secp256k1, if ECDSA is no longer safe due to quantum security, there are much bigger threat than the encapsulation described here.

**Review by xinbenlv — CHANGES_REQUESTED** [review]
These are my main editorial review comments, many of them are due to the grammar issue. I totally understand because I speak English as second language too. Please consider run via a grammar checker. 

Another major issue is to keep it consistent for terminologies that you specifically defined. For example, when addressing "Recipient" or "Sender", please capitalize.

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:33]: Many thanks for the very careful, very detailed review, and revision suggestions! @xinbenlv

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:116]: Good catch.

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md]: I'd like to retain this part as part of the justifications for the current design.

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md]: Good catch. Fixed and a commit pushed.

File: EIPS/eip-6051.md
Comment (weiji-cryptonatty) [reply EIPS/eip-6051.md:16]: I was just suggest to use `for` here. It looks like using `by` will give people an impression that the known or trusted party executes the action of the encapsulation.

**Comment by weiji-cryptonatty** [comment]
> These are my main editorial review comments, many of them are due to the grammar issue. I totally understand grammar challenge myself because I speak English as second language. Please consider run via a grammar checker.
> 
> Another major issue is to keep it consistent for terminologies that you specifically defined. For example, when addressing "Recipient" or "Sender", please capitalize.

Hi @xinbenlv thank you again for the very careful and very detailed reviews. I can see why EIP editors are so overwhelmed. I really appreciate what you had done here.

I think I have fixed all issues. I also ran another round with grammarly.com and fixed almost all issues it suggests (with very few I did not agree with it). Please review again when you do find time.

**Comment by xinbenlv** [comment]
From a technical perspective, I still think, if the EIP is defining a way to encapsulate and unwrap private key via `eth_` interface, that is considered an interface category, but we can address it later

**Review by xinbenlv — DISMISSED** [review]
I think this is good to merge as a draft now. Defer to @Pandapip1 @SamWilsn to follow up the merging.

**Comment by weiji-cryptonatty** [comment]
> I think this is good to merge as a draft now. Defer to @Pandapip1 @SamWilsn to follow up the merging.

File: EIPS/eip-6051.md
Comment (Pandapip1) [inline EIPS/eip-6051.md:9]:
```suggestion
category: Interface
```
This should be an Interface EIP, since it adds RPC endpoints.

File: EIPS/eip-6051.md
Comment (Pandapip1) [inline EIPS/eip-6051.md:438]:
```suggestion
```
This is not a reference implementation

File: EIPS/eip-6051.md
Comment (Pandapip1) [inline EIPS/eip-6051.md:442]: Please use headings to indicate headings, not bold text.

File: EIPS/eip-6051.md
Comment (Pandapip1) [inline EIPS/eip-6051.md:63]: Why not link to them?
```suggestion
With signature and verification, the signature to `R` by `singerPubKey` is appended to `R`. `signerPubKey` itself could have been already signed by `trustedPubKey`, and that signature is appended to `signerPubKey`. Note that the signature is applied to the byte array data instead of its string representation, which might lead to confusion and interoperability issues (such as hex or base64, lower case v.s. upper case, etc.). See [Requests](#requests) and [Test Cases](#test-cases) for further clarification and examples.
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

**Review by weiji-cryptonatty — COMMENTED** [review]
suggestions accepted

**Review by Pandapip1 — APPROVED** [review]
LGTM for a draft

**Pull request merged**
